AX: Set non-style-related attributes when serving attributed strings from the accessibility thread

twilco opened the pull request:
#### aff19f98fb652ba02cd995f7ee6b568864f5750f
<pre>
AX: Set non-style-related attributes when serving attributed strings from the accessibility thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=284117">https://bugs.webkit.org/show_bug.cgi?id=284117</a>
<a href="https://rdar.apple.com/140996511">rdar://140996511</a>

Reviewed by Chris Fleizach.

With this commit, we set many new attributes for attributed strings created lazily off the main-thread:
  - AXHighlight
  - AXIsSuggestedInsertion
  - AXIsSuggestedDeletion
  - AXIsSuggestion
  - AXLink
  - AXExpandedTextValue
  - AXHeadingLevel
  - AXBlockQuoteLevel
  - AXDidSpellCheck

After this commit, the two remaining attributes are related to eagerly resolving misspelling (rather than setting AXDidSpellCheck: 0),
and setting text autocompletion attributes (attributedStringSetCompositionAttributes), both of which currently require DOM nodes
or other main-thread state. We&apos;ll figure out what to do with these in a future patch.

Many new tests are added to verify we behave correctly.

* LayoutTests/accessibility/ax-thread-text-apis/attributed-string-includes-insertion-deletion-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string-includes-insertion-deletion.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/attributed-string-text-styling.html:
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/blockquote-level-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/blockquote-level.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/expanded-text-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/expanded-text.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/heading-level-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/heading-level.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/lazy-spellchecking-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/lazy-spellchecking.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/link-text-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/link-text.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/character-offset-from-upstream-position.html:
* LayoutTests/accessibility/ax-thread-text-apis/mark-role-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/mark-role.html: Added.
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::Accessibility::editableAncestor):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXTextMarker.cpp:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::attributedStringSetColor):
(WebCore::attributedStringSetExpandedText):
(WebCore::attributedStringSetBlockquoteLevel):
(WebCore::attributedStringSetNeedsSpellCheck):
(WebCore::attributedStringSetElement):
(WebCore::attributedStringSetStyle):
(WebCore::AXCoreObject::createAttributedString const):
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::attributedStringCreate):
(WebCore::attributedStringSetStyle): Deleted.
(WebCore::attributedStringSetHeadingLevel): Deleted.
(WebCore::attributedStringSetBlockquoteLevel): Deleted.
(WebCore::attributedStringSetExpandedText): Deleted.
(WebCore::attributedStringSetElement): Deleted.
(WebCore::attributedStringSetNeedsSpellCheck): Deleted.
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.h:

Canonical link: <a href="https://commits.webkit.org/287471@main">https://commits.webkit.org/287471@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bef15c19ef46e920b562541237e867d360a83287

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79782 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58783 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33181 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84307 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30789 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81891 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67857 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7074 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62360 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20201 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82850 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52416 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72657 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42664 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49758 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26805 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29232 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70887 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27276 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85733 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7017 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4914 "Found 1 new test failure: media/media-vp8-hiddenframes.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70617 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7190 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68496 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69856 "Found 5 new API test failures: /TestWebKit:WebKit2UserMessageRoundTripTest.WKURL, /TestWebKit:WebKit.LoadAlternateHTMLStringWithEmptyBaseURL, /TestWebKit:WebKit2.ProvisionalURLAfterWillSendRequestCallback, /TestWebKit:WebKit.FocusedFrameAfterCrash, /TestWebKit:WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginShouldNotBeDispatchedForAlreadyFocusedField (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17412 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13867 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12782 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6974 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12564 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6838 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10346 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8645 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on LayoutTests/accessibility/ax-thread-text-apis/attributed-string-includes-insertion-deletion.html:
```diff
@@ -0,0 +1,56 @@
+<!DOCTYPE HTML><!-- webkit-test-runner [ runSingly=true AccessibilityThreadTextApisEnabled=true ] -->
+<!-- This is a modern-style rewrite of existing test mac/mark-role.html. After accessibilityThreadTextApisEnabled is enabled by default, we should replace mac/mark-role.html with this file. -->
```
nit: `mac/mark-role.html` should be `mac/attributed-string-includes-insertion-deletion.html` here

twilco replied:
Good catch, thanks!

fleizach commented on LayoutTests/accessibility/ax-thread-text-apis/mark-role.html:
```diff
@@ -0,0 +1,40 @@
+<script>
+if (window.accessibilityController) {
+    window.jsTestIsAsync = true;
+    let output = "This test makes sure that the mark role exposes the right attributes in the attriubuted string.\n";
```
attriubuted -> attributed

twilco replied:
Fixed 👍

twilco committed:
AX: Set non-style-related attributes when serving attributed strings from the accessibility thread
https://bugs.webkit.org/show_bug.cgi?id=284117
rdar://140996511

Reviewed by Chris Fleizach.

With this commit, we set many new attributes for attributed strings created lazily off the main-thread:
  - AXHighlight
  - AXIsSuggestedInsertion
  - AXIsSuggestedDeletion
  - AXIsSuggestion
  - AXLink
  - AXExpandedTextValue
  - AXHeadingLevel
  - AXBlockQuoteLevel
  - AXDidSpellCheck

After this commit, the two remaining attributes are related to eagerly resolving misspelling (rather than setting AXDidSpellCheck: 0),
and setting text autocompletion attributes (attributedStringSetCompositionAttributes), both of which currently require DOM nodes
or other main-thread state. We'll figure out what to do with these in a future patch.

Many new tests are added to verify we behave correctly.

* LayoutTests/accessibility/ax-thread-text-apis/attributed-string-includes-insertion-deletion-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string-includes-insertion-deletion.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/attributed-string-text-styling.html:
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/blockquote-level-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/blockquote-level.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/expanded-text-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/expanded-text.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/heading-level-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/heading-level.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/lazy-spellchecking-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/lazy-spellchecking.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/link-text-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/attributed-string/link-text.html: Added.
* LayoutTests/accessibility/ax-thread-text-apis/character-offset-from-upstream-position.html:
* LayoutTests/accessibility/ax-thread-text-apis/mark-role-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/mark-role.html: Added.
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::Accessibility::editableAncestor):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXTextMarker.cpp:
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::attributedStringSetColor):
(WebCore::attributedStringSetExpandedText):
(WebCore::attributedStringSetBlockquoteLevel):
(WebCore::attributedStringSetNeedsSpellCheck):
(WebCore::attributedStringSetElement):
(WebCore::attributedStringSetStyle):
(WebCore::AXCoreObject::createAttributedString const):
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::attributedStringCreate):
(WebCore::attributedStringSetStyle): Deleted.
(WebCore::attributedStringSetHeadingLevel): Deleted.
(WebCore::attributedStringSetBlockquoteLevel): Deleted.
(WebCore::attributedStringSetExpandedText): Deleted.
(WebCore::attributedStringSetElement): Deleted.
(WebCore::attributedStringSetNeedsSpellCheck): Deleted.
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.h:

Canonical link: https://commits.webkit.org/287471@main
